parser/src/main/scala/org/bykn/bosatsu/parser/Parser.scala

```diff
@@ -0,0 +1,938 @@
+  }
+
+  sealed abstract class Error {
+    def expected: NonEmptyList[Expectation]
+    def offsets: NonEmptyList[Int] =
+      expected.map(_.offset).distinct
+
+    def translateOffset(inc: Int): Error =
+      this match {
+        case Error.MissedExpectation(x) => Error.MissedExpectation(x.translateOffset(inc))
+        case Error.Combined(es) => Error.Combined(es.map(_.translateOffset(inc)))
+      }
+  }
+  object Error {
+    case class MissedExpectation(expectation: Expectation) extends Error {
+      def expected = NonEmptyList(expectation, Nil)
+    }
+
+    case class Combined(errors: NonEmptyList[Error]) extends Error {
+      def expected = errors.flatMap(_.expected)
+    }
+
+    def combined(errors: NonEmptyList[Error]): Error =
+      if (errors.tail.isEmpty) errors.head
+      else Combined(errors)
+  }
+
+  final class With1[+A](val parser: Parser[A]) extends AnyVal {
+    def ~[B](that: Parser1[B]): Parser1[(A, B)] =
+      Parser.product01(parser, that)
+
+    /**
+     * This is the usual monadic composition, but you
+     * should much prefer to use ~ or Apply.product, *>, <*, etc
+     * if you can since it is much more efficient. This
+     * has to call fn on each parse, which could be a lot
+     * of extra work is you already know the result as is
+     * the case for ~
+     */
+     def flatMap[B](fn: A => Parser1[B]): Parser1[B] =
+        Parser.flatMap01(parser)(fn)
+  }
+
+  def pure[A](a: A): Parser[A] =
+    Impl.Pure(a)
+
+  /**
+   * this is an error if string is empty
+   */
+  def expect(str: String): Parser1[Unit] =
+    Impl.Expect(str)
+
+  def oneOf1[A](parsers: List[Parser1[A]]): Parser1[A] = {
+    @annotation.tailrec
+    def flatten(ls: List[Parser1[A]], acc: List[Parser1[A]]): Parser1[A] =
+      ls match {
+        case Nil =>
+          acc match {
+            case h :: Nil => h
+            case other => Impl.OneOf1(other.reverse)
+          }
+        case Impl.OneOf1(ps) :: rest =>
+          flatten(ps ::: rest, acc)
+        case notOneOf :: rest =>
+          flatten(rest, notOneOf :: acc)
+      }
+
+    flatten(parsers, Nil)
+  }
+
+  def oneOf[A](ps: List[Parser[A]]): Parser[A] = {
+    @annotation.tailrec
+    def flatten(ls: List[Parser[A]], acc: List[Parser[A]]): Parser[A] =
+      ls match {
+        case Nil =>
+          acc match {
+            case h :: Nil => h
+            case other => Impl.OneOf(other.reverse)
+          }
+        case Impl.OneOf(ps) :: rest =>
+          flatten(ps ::: rest, acc)
+        case notOneOf :: rest =>
+          flatten(rest, notOneOf :: acc)
+      }
+
+    flatten(ps, Nil)
+  }
+
+  /**
+   * Parse the next len characters where len > 0
+   */
+  def length(len: Int): Parser1[String] =
+    Impl.Length(len)
+
+  /**
+   * Repeat this parser 0 or more times
+   * note: this can wind up parsing nothing
+   */
+  def rep[A](p1: Parser1[A]): Parser[List[A]] =
+    Impl.Rep(p1)
```

Comment:
We could check that Rep isn't directly nesting. rep of rep won't parse any more, the inner one will scoop everything up.

This is especially useful potentially after a void.